src/riak_kv_query_server.erl

```diff
@@ -0,0 +1,678 @@
+    TimeoutS = riak_kv_query:get_timeout_secs(Query),
+    MaxResults = riak_kv_query:get_maxresults(Query),
+    AccType = accumulation_type(Query, MaxResults),
+    From = riak_kv_query:get_clientpid(Query),
+    ReqID = riak_kv_query:get_reqid(Query),
+    ClientMonitorRef = erlang:monitor(process, From),
+    case riak_core_coverage_plan:create_plan(all, NVal, R, ReqID, riak_kv) of
+        {error, Reason} ->
+            ?LOG_WARNING("Query coverage plan failed due to ~0p", [Reason]),
+            From ! {ReqID, {error, insufficient_vnodes}},
+            {stop, insufficient_vnodes};
+        {CoverageVnodes, FilterVnodes} ->
+            Sender = {raw, ReqID, self()},
+            Request =
+                riak_kv_requests:new_query_request(
+                    Bucket,
+                    none,
+                    riak_kv_query:get_querytype(Query),
+                    AccType,
+                    riak_kv_query:get_returnterms(Query),
+                    calculate_buffer_size(length(CoverageVnodes), MaxResults),
+                    EvaluatedQuery
+                ),
+            riak_core_vnode_master:coverage(
+                Request,
+                CoverageVnodes,
+                FilterVnodes,
+                Sender,
+                riak_kv_vnode_master
+            ),
+            InitMonitor = maps:from_keys(CoverageVnodes, 0),
+            VnodesOngoing = sets:from_list(CoverageVnodes, ?VERSION),
+            Acc =
+                case AccType of
+                    keys ->
+                        #list_acc{};
+                    raw_keys ->
+                        #list_acc{};
+                    terms ->
+                        #list_acc{};
+                    raw_terms ->
+                        #list_acc{};
+                    raw_count ->
+                        #count_acc{};
+                    count ->
+                        #count_acc{};
+                    term_with_rawcount ->
+                        #map_acc{};
+                    term_with_count ->
+                        #map_acc{}
+                    end,
+            erlang:send_after(TimeoutS * 1000, self(), {timeout, ReqID}),
+            {
+                ok, 
+                #state{
+                    from = {raw, ReqID, From},
+                    client_monitor = ClientMonitorRef,
+                    timeout_reqid = ReqID,
+                    req_id = ReqID,
+                    bucket = Bucket,
+                    vnode_monitor = InitMonitor,
+                    vnodes_ongoing = VnodesOngoing,
+                    acc = Acc,
+                    max_results = MaxResults,
+                    result_encoding_fun
+                        = riak_kv_query:get_result_encodingfun(Query)
+                }
+            }
+    end.
+
+handle_cast({new_timeout, SecondsToTimeout}, State) ->
+    ReqID = riak_kv_query:get_reqid(),
+    erlang:send_after(SecondsToTimeout * 1000, self(), {timeout, ReqID}),
```

Comment:
This seems tricky.
You compute a new ReqId for the new timeout, but you don't clear the ongoing timeout. That's a nice way to avoid a race.

However, in the case your old timer goes of, I think there still is a message coming in `{timeout, OldReqId}`. But that will not match on line 253 below. Hence it falls all the way through and the catch all case stops the server, which is likely not what you intended.

I suggest a second handle info to ingnore timeouts that have the wrong ReqId